### ISSUE TITLE
Add no-unreachable-ternary

### ISSUE THREAD

**Issue by GrayedFox** [eslint/eslint#13754]
**Please describe what the rule should do:**
This rule will detect ternary conditions that necessarily result in unreachable code (except for when the conditions have side effects). It will also detect unreachable code by inverting any nested conditional expression to see if it is an "inverted duplicate" as well as catching equivalent OR expressions, i.e. `x || y` is logically equivalent to `y || x`.

**What new ECMAScript feature does this rule relate to?**
None

**What category of rule is this? (place an "X" next to just one item)**
- [x] Warns about a potential error (problem)

**Provide 2-3 code examples that this rule will warn about:**
```js
const getFee = (user) => user.isMember ? user.isMember ? 2.00 : 3.00 : 4.00;
// 3.00 is unreachable

const thing = condition1 || condition2 ? condition2 || condition1 ? 'x' : 'y' : 'z';
// 'y' is unreachable

const getFee = (user) => user.isMember ? 3.00 : user.isMember ? 2.00 : 4.00;
// 2.00 is unreachable

const result = foo(x) ? !foo(x) ? 1 : 2 : 3;
// 1 is unreachable

const formula = condition1 && condition2
    ? condition2
    ? `${oldRate}+${User.fee}`
    : `${currentRate}+${User.fee}`
    : `${newRate}+${User.fee}`;
// `${currentRate}-${User.fee} is unreachable`

condition1 ? foo() : condition2 ? bar() : condition3 ? baz() : condition1 ? qux() : bar());
// qux() is unreachable
```

**Why should this rule be included in ESLint (instead of a plugin)?**
Because this pattern nearly always results in unreachable code which is most likely unintended by the developer, thus the rule is also applicable to any/all projects.

**Are you willing to submit a pull request to implement this rule?**
Yes!

Extra: this behaviour was first discussed [here](https://github.com/eslint/eslint/issues/13744) but the general direction seems to be in favour of splitting the rule into distinct parts, `no-dupe-ternary-expressions` and `no-unreachable-ternary` (although consensus is yet to be reached).

**Comment by nzakas** [comment]
For the same [reasons](https://github.com/eslint/eslint/issues/13744#issuecomment-708088991 ) as #13744, I’m not in favor of accepting this as a core rule.

**Comment by GrayedFox** [comment]
Will likely turn this into a plug in over the next days - feel free to close if consensus has not been reached (any review is still of course very welcome!) 🙇🏾  #13766
